Tooling with bnd, Http-Whiteboard Example

stbischof opened the pull request:
Tooling with bnd, Http-Whiteboard

- build, integration-test, export and run with bnd
- using Http-Whiteboard-Annotations in `better` project
- bnd-reporter for readme-documentation



<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/base-starter-flow-osgi/94)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/base-starter-flow-osgi?pullRequest=94) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/base-starter-flow-osgi?pullRequest=94) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/base-starter-flow-osgi?pullRequest=94) it.</sub>

stbischof committed:
Tooling with bnd, Http-Whiteboard
- build, integration-test, export and run with bnd
- using Http-Whiteboard-Annotations in `better` project
- bnd-reporter for readme-documentation

Signed-off-by: Stefan Bischof <stbischof@bipolis.org>

stbischof committed:
document hot deployment

Signed-off-by: Stefan Bischof <stbischof@bipolis.org>

denis-anisimov commented:
Thank you for your contribution.
I hope Pekka mentioned to you that we won't merge this PR as is but instead will use it for investigation and making our own PR to follow similar way.

I'm going to ask you questions about some things here since some of them simply don't work or I was not able to find documentation about them.

denis-anisimov commented on README.md:
```diff
@@ -7,10 +7,30 @@ The best way to use it is via [vaadin.com/start](https://vaadin.com/start) - you
-This command will run profiles which downloads for you Felix OSGi container and installs `felix-jetty` 
-as OSGi web container. It deploys all necessary bundles to Felix and start it synchronously.
-The container will be stopped  when you stop your maven build (e.g. by killing it via `Ctrl+C`).
+Main Command: `mvn clean install`
```
This command just fails for me.

But may be it doesn't really matter since I'm going to create another PR

stbischof replied:
it fails on an [integration-test](https://github.com/vaadin/base-starter-flow-osgi/blob/5f78e2fe208a36816369e51ea6f0386190a080fe/starter.flow.better/src/test/java/com/example/starter/flow/better/Test.java) 

you may have changed to much. i am absolutely sure that this builds and tests work with older version

denis-anisimov replied:
`Lookup.lookup(ResourceProvider.class)` has to return not null object otherwise OSGi integration is simply totally broken.

Even though `Lookup.lookup` may be changed to `Optional`  but this is a different question and most likely it will complicate the usage.

denis-anisimov replied:
Our IT tests work fine : `ResourceProvider` instance is available in lookup.

I had similar issue in our IT tests but I was not able to figure out what's going on there.
Apparently it's related to the fact _HOW_ the servlet is registered.
In case of bundle Activator : https://github.com/vaadin/flow/blob/2.4/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/Activator.java , 
https://github.com/vaadin/flow/blob/2.4/flow-tests/test-root-context/bnd.bnd#L9
then service https://github.com/vaadin/flow/blob/feature/osgi/flow-osgi/src/main/java/com/vaadin/flow/osgi/support/OSGiResourceProvider.java#L41 is not yet available at the servlet registration point.

But if I register a servlet this way: https://github.com/vaadin/flow/blob/feature/osgi/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/Activator.java#L48
then everything works fine.

So it's some race condition which is most likely also the case for you as well.

May be you know how to make this work properly ?
May be resource provider should be registered via a Bundle activator and not via Declarative service.

This 
```
@Reference
	ResourceProvider pr;
```
is more a workaround and not really a solution: you should not know anything about `ResourceProvider` which is internal SPI.

stbischof replied:
i know that this is just an workaround! but it helps to show what the issue is.

This is an often as question: [best awnser](https://stackoverflow.com/questions/32455624/order-of-deployment-of-bundles)

You simply can not expect that a bundle/ service is exists. If you uninstall a bundle A1 with a service S1 and install a new version A2  at runtime all services of bundle B that needs this S1 have must be deactivated on uninstall and activated again on install.

DS will do all this for you.

denis-anisimov replied:
Well, it relates to the order and it doesn't.

I don't rely on any order when I ask for `ResourceProvider`. 
The service is asked at runtime directly : https://github.com/vaadin/flow/blob/feature/osgi/flow-server/src/main/java/com/vaadin/flow/server/osgi/OSGiAccess.java#L95 on demand.

So there is no "programmatic" issues with that.

But the issue is : the servlet is registered too early when `ResourceProvider`  is not yet available.
Since the developer doesn't not anything about `ResourceProvider`  internal SPI it becomes an issue.

Please note: _this is not an issue of bundle deployment order_.
This is an issue with a servlet registration.

Once we implement an automatic servlet registration we will just force the waiting for the `ResourceProvider` service availability. 
But for now when developer has to register the servlet himself (and we will allow to do this anyway even with automatic servlet registration) we have to provide some mechanism which register a servlet only when there is already `ResourceProvider`. And I don't see how to do this properly.

`flow-osgi`  bundle is required for ANY Vaadin WAB. So it has to be deployed before the WAB.
Yes, it may be redeployed at any time but  the service should be available at the servlet registration only.
A Vaadin WAB may not be activated before its dependencies are activated. `flow-osgi`  is one of such dependencies.
So the activation of WAB should happen after the activation of `flow-osgi`.
So I think I may rely on the fact that a bundle activator of WAB is called (or any declarative service is registered) _after_ the `flow-osgi`  bundle Activator is called. 
So I think I can just move the service registration in the bundle Activator and it will just work.

Another option is to move the OSGi logic implemented in the https://github.com/vaadin/flow/blob/feature/osgi/flow-osgi/src/main/java/com/vaadin/flow/osgi/support/OSGiResourceProvider.java#L40 into `flow-server`  module but I don't like this way.

stbischof replied:
I agree!
>> Please note: this is not an issue of bundle deployment order.
>> This is an issue with a servlet registration.

but most will try to fix this using the bundle deployment order. that would be wrong!

>> Another option is to move the OSGi logic implemented in the https://github.com/vaadin/flow/blob/feature/osgi/flow-osgi/src/main/java/com/vaadin/flow/osgi/support/OSGiResourceProvider.java#L40 into flow-server module but I don't like this way.

was my first idea. not sure that this is good.

>> Once we implement an automatic servlet registration we will just force the waiting for the ResourceProvider service availability.
But for now when developer has to register the servlet himself (and we will allow to do this anyway even with automatic servlet registration) we have to provide some mechanism which register a servlet only when there is already ResourceProvider. And I don't see how to do this properly.

I have to understand more why you use this ResourceProvider at the moment of activating the service

denis-anisimov replied:
This is an abstraction for strategy to access classpath resources.
In OSGi the resources can't be access the same way as in non-OSGi where resource is simply always in the classpath of the web app and you may access it via `SomeClass.class.getResource`.
In OSGi the correct way is `Bundle.getResource`  (and the Bundle here is not the same bundle as `flow-server`  bundle). So the resource access strategy is extracted to an abstraction layer and two implementations are provided:
for OSGi and non OSGi case.

denis-anisimov replied:
I would like to extract OSGi related code as much as possible to `flow-osgi` and don't have anything related to OSGi in `flow-server` or have a minimum things there.

denis-anisimov commented on README.md:
```diff
@@ -7,10 +7,30 @@ The best way to use it is via [vaadin.com/start](https://vaadin.com/start) - you
+- run tests inside the osgi framework when `tests.bndrun` file exists in project.
+- generates the Application if `application.bndrun` file exists in project.
+
+Run the Application using `java -jar application/application.jar`
```
Since `install` command fails `application.jar` is not generated.

stbischof replied:
works now! I fixed the path in the docs.  added (`target`) folder
`java -jar application/target/application.jar`

denis-anisimov replied:
The main reason why it failed for me is: `httpclient-4.5.13.jar` is included in the final application jar instead of the correct one `httpclient-osgi.jar`.

You have excluded explicitly  the artifact 
```
<groupId>org.apache.httpcomponents</groupId>
<artifactId>httpclient</artifactId>
```
from the build.
But :
- this is not required since this artifact is not an OSGi bundle ( so it won't affect anything )
- it doesn't help to me anyhow: the artifact is still included
- what's more important the valid OSGi bundle `httpclient-osgi.jar` is never included into the produced application jar.

As a result : the app doesn't work just because main `flow-server`  bundle dependencies are not installed , as a result `flow-server`  is not active and nothing works in OSGi at all.

I have not checked your latest changes but I'm pretty sure that it has not improved anyhow.
In my project which I create from scratch following your way the issue is exactly the same.

It seems that `resolve`  goal is not able to resolve the artifact properly because both jars `httpclient-osgi.jar` and `httpclient.jar` contains the same packages and `bnd-resolver:resolve` simply fails to select the correct jar.

I'm trying to resolve this issue now.

stbischof replied:
https://github.com/bndtools/bnd/issues/4380

denis-anisimov replied:
I've made httpclient bundle optional in Flow.
Now I'm able to build and run the app without unsatisfied dependencies.

But HTTP doesn't work : https://github.com/vaadin/base-starter-flow-osgi/pull/97

stbischof replied:
HTTP Whiteboard is working
http://localhost:8080/icons/icon.png


But VaadinServlet.service has a problem
```
javax.servlet.ServletException: com.vaadin.flow.server.ServiceException: java.lang.NullPointerException
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:254)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.apache.felix.http.base.internal.handler.ServletHandler.handle(ServletHandler.java:123)
	at org.apache.felix.http.base.internal.dispatch.InvocationChain.doFilter(InvocationChain.java:86)
	at org.apache.felix.http.base.internal.dispatch.Dispatcher$1.doFilter(Dispatcher.java:146)
	at org.apache.felix.http.base.internal.whiteboard.WhiteboardManager.invokePreprocessors(WhiteboardManager.java:988)
	at org.apache.felix.http.base.internal.dispatch.Dispatcher.dispatch(Dispatcher.java:91)
	at org.apache.felix.http.base.internal.dispatch.DispatcherServlet.service(DispatcherServlet.java:49)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:569)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1377)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:507)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1292)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:501)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.vaadin.flow.server.ServiceException: java.lang.NullPointerException
	at com.vaadin.flow.server.VaadinService.handleExceptionDuringRequest(VaadinService.java:1617)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1580)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:252)
	... 33 more
Caused by: java.lang.NullPointerException
	at com.vaadin.flow.server.frontend.FrontendUtils.getResourceFromWebpack(FrontendUtils.java:420)
	at com.vaadin.flow.server.frontend.FrontendUtils.getStatsAssetsByChunkName(FrontendUtils.java:528)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.appendNpmBundle(BootstrapHandler.java:904)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupFrameworkLibraries(BootstrapHandler.java:886)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.setupDocumentHead(BootstrapHandler.java:765)
	at com.vaadin.flow.server.BootstrapHandler$BootstrapPageBuilder.getBootstrapPage(BootstrapHandler.java:539)
	at com.vaadin.flow.server.BootstrapHandler.synchronizedHandleRequest(BootstrapHan
```

denis-anisimov replied:
>HTTP Whiteboard is working
>http://localhost:8080/icons/icon.png

True.
The icon is shown.


>But VaadinServlet.service has a problem

My bad.

To be able to run project correctly one has to :
- go to the `starter`  module and run : `mvn install -Pproduction`. `production` profile is important because it runs webpack which creates the stats file and then there will be no the exception above.
- then go to the `app`  and run `mvn install`.
- then run the app from this module as `java -jar target/app.jar`.

I don't see any exception when I access `http://localhost:8080` but nothing is shown. The page gives me 404.

stbischof replied:
ah production mone.. ok!

Use this:
```
mvn install -Pproduction
mvn bnd-resolver:resolve -f app/pom.xml 
mvn bnd-run:run -f app/pom.xml 

```

stbischof replied:
Now i tried it also with your VaadinServletRegistration.

this is only working when

- adding a reference to ResourceProvider

```
@Reference(cardinality = ReferenceCardinality.MANDATORY)
ResourceProvider pr;
```

- chnng bundle order after resolving

denis-anisimov replied:
Hmmm....
Interesting.
So it works for you...

It seems there should be some exceptions.
But I don't see any.
And 404 should not be returned on this case. But rather 503.
I really dislike how things work here: no any way to understand what's wrong.
But I will check tomorrow how it behaves with a reference to the `ResourceProvider`.

Anyway: I have to say thank you very much for you help even though I still have problems your help is very valuable !

stbischof replied:
ah sorry i had this .
i had this in my base. 

```
-runrequires.flow: \
	bnd.identity;version='2.5.0';id='com.vaadin.flow.client',\
	bnd.identity;version='2.5.0';id='com.vaadin.flow.data',\
	bnd.identity;version='2.5.0';id='com.vaadin.flow.html.components',\
	bnd.identity;version='2.5.0';id='com.vaadin.flow.push'
```

this is what runs in my computer.
https://github.com/bipolis/base-starter-flow-osgi/tree/feature/osgi-bnd

Changes 
https://github.com/bipolis/base-starter-flow-osgi/commit/801961abcc10bbc8bf14cdaa16b5e4e9f445fd61


Not all nessecary bundles are chatched from resolver.

stbischof replied:
Prevent activation of a VaadinServlet before ResourceProvider exists.
Q: What starts the activation of a VaadinServlet?
A: VaadinServletRegistration is `immediate=true`. - will activate on bundle-start
So i removed VaadinServletRegistration and used only the FixedVaadinServlet. `immediate=false` by default.

Q: What activate the FixedVaadinServlet
A: HttpWhiteboard looks for every Servlet and activates, if the Servlet matches the Servletcontext. This is when the Bundle of the FixedVaadinServlet is started AND the Whiteboards will be initialized. does not help enough- could be before ResourceProvider is started.

So i bound the FixedVaadinServlet to an other new ServletContext then the Default one.
https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.http.whiteboard.html#service.http.whiteboard.common.properties
```
@Component(service = { Servlet.class
		// Same like @VaadinContext
		// , property = {HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_SELECT + "=(" +
		// HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_NAME +"=vaadin")"})

})
@VaadinMode
@VaadinContext
@HttpWhiteboardServletAsyncSupported
@HttpWhiteboardServletPattern("/*")
public class FixedVaadinServlet extends VaadinServlet {

```
Now this Servlet would not be activates from HttpWhiteboard with the DefaultContext

Log:
```
[DEBUG] [ServiceReference 30 from bundle 2 : com.example.osgi.base:1.0.0 ref=[javax.servlet.Servlet] properties={component.id=0, component.name=com.example.starter.base.osgi.FixedVaadinServlet, objectClass=[javax.servlet.Servlet], osgi.http.whiteboard.context.select=(osgi.http.whiteboard.context.name=vaadin), osgi.http.whiteboard.servlet.asyncSupported=true, osgi.http.whiteboard.servlet.pattern=[/*], service.bundleid=2, service.id=30, service.scope=bundle, servlet.init.compatibilityMode=false}] Ignoring unmatching Servlet service

```
The FixedVaadinServlet will now ne activated in the moment when an ServletContext with the name `vaadin` exists.
ServletContexts couls be created using a ServletContextHelper.

```
@Component(service = ServletContextHelper.class, scope = ServiceScope.BUNDLE, property = {
		HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_NAME + "=" + VaadinServletContext.NAME,
		HttpWhiteboardConstants.HTTP_WHITEBOARD_CONTEXT_PATH + "=/" })
public class VaadinServletContext extends ServletContextHelper {

	@Reference //maybe not longer needed  when places in flow-osgi bundle
	ResourceProvider pr;
	public static final String NAME = "vaadin";
}
```
IF we put this ServletContextHelper into the flow-osgi bundle the VaadinServletContext would be activated by the whiteboard AFTER the immediate activation of theOSGiResourceProvider.

at the moment i just testet the VaadinServletContext inside the starter bundle. but i am quiet sure this could work!

I created the annotation @VaadinContext. this creates ComponentPropertys that prevents from starting with default ServletContext.

denis-anisimov replied:
Interesting solution with `VaadinContext`. 
It might be that we will have to use it later on, but :
- I'm pretty sure that I will find a solution with `ResourceProvider` without this.
- that an additional annotation which one has to know: it means that simplest way of a plain servlet registration won't work. It seems this is a big limitation just to solve one `ResourceProvider` issue. But let's see.

I would like to return back to the my `osgi-bnd`  branch:
thanks a lot for your changes. Now I'm able to understand what's going on there.

What I've got finally:
`-runrequires.flow` - this section will be merged into `-runrequires`. It's not referenced from anywhere. It's just a convenience way to define run requirements instead of having them all together in one `-runrequires`. 
It's just the same as having one `-runrequires` which has all the lines from all the `-runrequires.xxx`. 

That what I was not able to understand and now it's clear.
`-runrequires.flow`  declares additional bundles which `resolve`  was not able to discover as necessary dependencies and we have to declare them explicitly to deploy because without them a Vaadin WAB may not work.

That was the first reason why I've got 404 .

Another reason is `ResourceProvider`  issue.
Now it becomes important.
Indeed if I add 
```
@Reference(cardinality = ReferenceCardinality.MANDATORY)
ResourceProvider pr;
```
then I see a button in the browser.

Without this 404 is shown.

**BUT** the extremely important question is : why it's 404? Why it's not 503 ?
Apparently the only reason is : some exception which is thrown because `ResourceProvider` service is not yet available.
I don't see any exception in the console and 404 instead of 503 is returned.
This is quite severe issue because there is no any way to understand what's wrong : no any error description, no any exception , it just doesn't work.
I strongly dislike this magic.
Do you know why it happens ? May be there is some log which I may check and where the exception is available ?

stbischof replied:
>> That what I was not able to understand and now it's clear.
>> -runrequires.flow declares additional bundles which resolve was not able to discover as necessary dependencies and we >> have to declare them explicitly to deploy because without them a Vaadin WAB may not work.

There would also be a way to have an Annnotation @RequireVaddin (maybe implicet in @VaadinContext) which adds a requirement to the bundle manifest.

And the Capabilities would be provided from vaadin server bundles within the manifest.

That is the same Mechanism why you get a the HTTP-Whiteboard Implementattion while resolving only be adding @HttpWhiteboardServletPattern("/game") to a Servlet.

denis-anisimov commented on README.md:
```diff
@@ -7,10 +7,30 @@ The best way to use it is via [vaadin.com/start](https://vaadin.com/start) - you
+
+
+Special Commands:
+- Resolve the Framework using `mvn bnd-resolver:resolve application/pom.xml`
```
This command simply fails just because 

```
No plugin found for prefix 'bnd-resolver' in the current project and in the plugin groups
```

stbischof replied:
mvn bnd-resolver:resolve -f application/pom.xml

denis-anisimov replied:
Unfortunately this goal is quite cryptic: the sentence "it resolves" doesn't say anything to the end user.

It resolves: yes. But for me it meant that the bundles are printed in the console and that's it.
Apparently :
- if I have `-runbundles:` in the bndrun file then the resolution happens temporary only: it prints the resolved bundles in the log and that's it.
- I had to remove `-runbundles:` from bndrun file completely and only after that I see that the goal has written the resolved bundles into the bndrun.

This is extremely confusing and undocumented behavior :(

denis-anisimov commented on application/application.bndrun:
```diff
@@ -0,0 +1,58 @@
+-include: ../base.bndrun
+
+
+-runrequires.webconsole: \
```
I was not able to find in any documentation `-runrequires.webconsole` notation.

There is a documentation about root bundle in `-runrequires`.

But there is nothing about prefix `-runrequires` , so I don't understand what `-runrequires.webconsole` is for.
Could you please explain this ?

stbischof replied:
the `-runrequires` [documentation](https://bnd.bndtools.org/instructions/runrequires.html) holds filters to all main bundles that you needed. if you hit the resolve button or start resolving in maven it will generate the -runbundles automatically and finds all necessary dependencys.

so `-runbundles` could be removed from from this file because they will resolves via the resolve plugin. I often let them in the file to know what bundles changes (also an fail on change option in the plugin).

the suffix `.webconsole` or `.flow` is just to group the required bundles ans give the group a name.

denis-anisimov replied:
Well, I assumed that `.webconsole` is an identifier of something similar to `-runrequires`.
But what I don't understand is: how it's referenced ?

If I remove `-runrequires.flow`  then it seems nothing is broken. So these groups are never used ?
So why they are defined at all then ?
Apparently this is not a "reserved keyword" for bnd tool. So it has to be used in the config somewhere.
But I don't see this usage.

denis-anisimov commented on application/application.bndrun:
```diff
@@ -0,0 +1,58 @@
+-runrequires.webconsole: \
+	bnd.identity;id='org.apache.felix.webconsole'
+
+-runrequires.flow: \
```
Same as above: I don't understand what this is for.

stbischof replied:
in the application l additionally would have the felix webconsole so see state of services/bundles/framework.

stbischof committed:
fixed

Signed-off-by: Stefan Bischof <stbischof@bipolis.org>

stbischof committed:
auto-resolve application

Signed-off-by: Stefan Bischof <stbischof@bipolis.org>

stbischof commented:
now it runs again after all your changes. Tell me if you have more questions on the OSGi / bnd topic.

stbischof commented:
Maybe we should talk on gitter/ webchat. That would may be easier.

denis-anisimov commented on starter.flow.better/src/main/java/com/example/starter/flow/better/osgi/VaadinMode.java:
```diff
@@ -0,0 +1,40 @@
+@RequireHttpWhiteboard
+public @interface VaadinMode {
+
+	String PREFIX_ = "servlet.init.";
```
The constant is not used from anywhere.

Is it some magic ?
Is it documented somewhere ?

pleku commented:
Closing as this is outdated and we are not working anymore with the targeted branch.
